### PR TITLE
CONTEXT: lazily evaluate frames to only actually assemble when needed

### PR DESCRIPTION
**Pull request by Al2Klimov**
```
[2022-10-28 17:38:10 +0200] information/cli: Icinga application loader (version: v2.13.0-461-g8e3a13020; debug)
[2022-10-28 17:38:10 +0200] information/cli: Loading configuration file(s).
[2022-10-28 17:38:11 +0200] information/ConfigItem: Committing config item(s).
[2022-10-28 17:38:11 +0200] critical/config: Error: Error while evaluating expression: Tried to access undefined script variable 'x'
Location: in /Users/aklimov/NET/WS/icinga2/prefix/etc/icinga2/conf.d/services.conf: 31:53-31:53
/Users/aklimov/NET/WS/icinga2/prefix/etc/icinga2/conf.d/services.conf(29):   check_command = "ping4"
/Users/aklimov/NET/WS/icinga2/prefix/etc/icinga2/conf.d/services.conf(30):
/Users/aklimov/NET/WS/icinga2/prefix/etc/icinga2/conf.d/services.conf(31):   assign where host.address && ((function() { throw x })())
                                                                                                                               ^
/Users/aklimov/NET/WS/icinga2/prefix/etc/icinga2/conf.d/services.conf(32): }
/Users/aklimov/NET/WS/icinga2/prefix/etc/icinga2/conf.d/services.conf(33):

Context:

	(0) Evaluating 'apply' rule (in /Users/aklimov/NET/WS/icinga2/prefix/etc/icinga2/conf.d/services.conf: 26:1-26:21)
	(1) Evaluating 'apply' rules for host 'alexandersmbp2.int.netways.de'
[2022-10-28 17:38:11 +0200] critical/config: 1 error
[2022-10-28 17:38:11 +0200] critical/cli: Config validation failed. Re-run with 'icinga2 daemon -C' after fixing the config.
```

closes #9556

### CODE REVIEW

File: lib/base/context.hpp [hunk 1]
```diff
@@ -33,18 +34,21 @@ std::ostream& operator<<(std::ostream& stream, const ContextTrace& trace);
...
 
 	friend class ContextTrace;
 };
 
 /* The currentContextFrame variable has to be volatile in order to prevent
  * the compiler from optimizing it away. */
-#define CONTEXT(message) volatile icinga::ContextFrame currentContextFrame(message)
+#define CONTEXT(message) volatile icinga::ContextFrame currentContextFrame ([&](std::ostream& stream) { \
+stream << message; \
+})
```
Comment (julianbrost) [inline lib/base/context.hpp:50]: This change means that the caller has to ensure that whatever expression is used leads to a constant result while the scope is active. Not impossible but requires review of all callers.

**Review by julianbrost — COMMENTED** [review]
You have already changed some callers of `CONTEXT()`. There are quite a few that use the `+` operator for building the string. It might make sense to consistently use one or the other (less from a performance point of view but more from a "this is the intended way to use this" point of view).

File: lib/base/context.hpp [hunk 2]
```diff
@@ -33,18 +34,21 @@ std::ostream& operator<<(std::ostream& stream, const ContextTrace& trace);
...
+	static std::vector<std::function<void(std::ostream&)>>& GetFrames();
 
 	friend class ContextTrace;
 };
 
 /* The currentContextFrame variable has to be volatile in order to prevent
  * the compiler from optimizing it away. */
-#define CONTEXT(message) volatile icinga::ContextFrame currentContextFrame(message)
+#define CONTEXT(message) volatile icinga::ContextFrame currentContextFrame ([&](std::ostream& db026162_4822_4144_95e0_4a5a42d1c1b3) { \
+db026162_4822_4144_95e0_4a5a42d1c1b3 << message; \
```
Comment (julianbrost) [inline lib/base/context.hpp:49]: That name is really not helpful at all.

Either go with something that is unlikely to conflict but still makes sense as a name (something like `_CONTEXT_stream`) or, if you want to remove all potential conflicts, go with the other version I suggested:

```c++
#define CONTEXT(message) volatile icinga::ContextFrame currentContextFrame ([&]() { \
    return (::std::stringstream() << message).str(); })
```

File: lib/base/context.cpp [hunk 3]
```diff
@@ -3,32 +3,40 @@
...
 
 ContextTrace::ContextTrace()
-	: m_Frames(ContextFrame::GetFrames())
-{ }
+{
+	for (auto frame (ContextFrame::GetFrames().rbegin()); frame != ContextFrame::GetFrames().rend(); ++frame) {
+		std::ostringstream oss;
+
+		(*frame)(oss);
+		m_Frames.emplace_back(oss.str());
```
Comment (julianbrost) [inline lib/base/context.cpp:37]: ... then you could just do `m_Frames.emplace_back((*frame)());` here, which wouldn't even be any more expensive compared to what you're doing here at the moment (and even if it was, it wouldn't matter, when this is executed, we're on a slow path already anyways).

**Review by julianbrost — CHANGES_REQUESTED** [review]

File: lib/base/context.cpp (hunk 3 above)
Comment (Al2Klimov) [reply lib/base/context.cpp:37]: What for?

Comment (julianbrost) [reply lib/base/context.cpp:37]: This comment was a continuation of https://github.com/Icinga/icinga2/pull/9557#discussion_r1034818584. Changing the lambda to return a string would just have allowed simplifying the code here a bit as well.

**Review by julianbrost — APPROVED** [review]

**Pull request merged**